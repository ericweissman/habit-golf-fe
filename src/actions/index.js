export const isLoading = (bool) => ({
  type: 'IS_LOADING',
  isLoading: bool
})

export const hasErrored = (error) => ({
  type: 'HAS_ERRORED',
  error
})

export const loginUserSuccess = (user) => ({
  type: 'LOGIN_USER_SUCCESS',
  user
})

export const setActiveClub = (club) => ({
  type: 'SET_ACTIVE_CLUB',
  club
})

export const setLastShot = (shot) => ({
  type: 'SET_LAST_SHOT',
  shot
})