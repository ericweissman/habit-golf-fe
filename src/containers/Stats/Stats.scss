.stats {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.graph {
  width: 260px;
}

.date-btns {
  @extend .club-btns;
  width: 200px;
  margin: 20px;
  .active-true {
    color: $masters-yellow;
    font-weight: 700;
  }
}

.stats-breakdown {
  background-color: $masters-green;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
  margin-top: -110px;
  border-radius: 10px;
  width: 270px;
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  text-align: center;
  font-size: 28px;
  padding: 10px 0px;
  grid-gap: 5px;
  p {
    color: white;
    font-weight: 200;
    padding: 5px 0;
  }
  span {
    font-size: 18px;
    margin-left: 5px;
  }
}

.back-btn {
  @extend .take-shot-btn;
  width: 100px;
  height: 40px;
  margin-top: 15px;
}