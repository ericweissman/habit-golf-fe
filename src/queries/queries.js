export const postShotData = (clubId, rating, id) => ({
  query: `mutation {
      createShot( clubId: ${clubId}, rating: ${rating}, playerId: ${id}  ) {
        rating
        clubId
        playerId
    }
  }`
})

export const getShotDataAllTime = (id) => ({
  query: `{
    playerStats(playerId: ${id}){
      playerId
      greatShotPercentage
      hookPercentage
      pullPercentage
      pushPercentage
      slicePercentage
      duffPercentage
    }
  }`
})

export const getShotDataToday = (id) => ({
  query: `{
    playerStats(playerId: ${id}){
      playerId
      todayGreatShotPercentage
      todayHookPercentage
      todayPullPercentage
      todayPushPercentage
      todaySlicePercentage
      todayDuffPercentage
    }
  }`
})