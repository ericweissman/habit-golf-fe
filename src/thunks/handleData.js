import { isLoading, hasErrored } from '../actions'

export const handleData = (actionToDispatch, query) => {
  const uri = "https://habit-golf-api.herokuapp.com/graphql"
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(uri, {
        method: "POST",
        headers: {
          "Content-Type": "application/json"
        },
        body: JSON.stringify(query)
      })
      if(!response.ok) {
        throw Error(response.errors[0].message)
      }
      dispatch(isLoading(false))
      const result = await response.json()
      dispatch(actionToDispatch(result))
    } catch (error) {
      dispatch(hasErrored(error.message))
      dispatch(isLoading(false))
    }
  }
}