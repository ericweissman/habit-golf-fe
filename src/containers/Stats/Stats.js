import React, { Component } from 'react';
import PieChart from 'react-minimal-pie-chart';
import { connect } from 'react-redux';
import { handleData } from '../../thunks/handleData'
import { getShotData } from '../../queries/queries'
import { getDataSuccess } from '../../actions'
import Loading from '../../components/Loading/Loading'


export class Stats extends Component {
  state = {
    range: 'today',
  }

  componentDidMount() {
    this.props.handleData(getDataSuccess, getShotData(1))
  }

  changeRange = (e) => {
    const { value } = e.target
    this.setState({ range: value })
  }

  render() {
    const { changeFilter, isLoading } = this.props
    const { greatShotPercentage, hookPercentage, pullPercentage, pushPercentage, slicePercentage } = this.props.shotData
    const data = [
      { title: 'Hook', value: hookPercentage, color: '#ca0032', },
      { title: 'Pull', value: pullPercentage, color: '#FEE715' },
      { title: 'Great', value: greatShotPercentage, color: '#076652' },
      { title: 'Push', value: pushPercentage, color: '#FEE715' },
      { title: 'Slice', value: slicePercentage, color: '#ca0032' }]
    return (
        isLoading ? <Loading type="bubbles"/> : 
        <div className='stats'>
          <div className="date-btns">
            <button className={'active-' + (this.state.range === 'today')} onClick={this.changeRange} value="today">today</button>
            <button className={'active-' + (this.state.range === 'all time')} onClick={this.changeRange} value="all time">all time</button>
          </div>
          <div className="graph-area">
            <PieChart
              className="graph"
              data={data}
              startAngle={180}
              lengthAngle={180}
              animate
            />
          </div>
          <div className='stats-breakdown'>
            <p>hook <span>{hookPercentage}%</span></p>
            <p>great <span>{greatShotPercentage}%</span></p>
            <p>slice <span>{slicePercentage}%</span></p>
            <p>pull <span>{pullPercentage}%</span></p>
            <p>duff <span>{slicePercentage}%</span></p>
            <p>push <span>{pushPercentage}%</span></p>
          </div>
          <button className="back-btn" onClick={() => changeFilter('home')}>Back</button>
        </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  shotData: state.shotData,
  isLoading: state.isLoading,
})
export const mapDispatchToProps = (dispatch) => ({
  handleData: (actionToDispatch, query) => dispatch(handleData(actionToDispatch, query))
})

export default connect(mapStateToProps, mapDispatchToProps)(Stats);
