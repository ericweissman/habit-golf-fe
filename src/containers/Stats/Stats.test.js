import React from 'react';
import { Stats, mapDispatchToProps, mapStateToProps } from './Stats';
import { handleData } from '../../thunks/handleData'
import { getShotDataAllTime, getShotDataToday } from '../../queries/queries'
import { getDataSuccess } from '../../actions'
import { shallow } from 'enzyme'

jest.mock('../../thunks/handleData')

describe('Stats', () => {
  let wrapper
  const mockProps = {
    shotData: {
      greatShotPercentage: 10,
      hookShotPercentage: 10,
      duffShotPercentage: 10,
      sliceShotPercentage: 10,
      pushShotPercentage: 10,
      pullShotPercentage: 10,
    },
    handleData: jest.fn(),
    isLoading: false,
    changeFilter: jest.fn()
  }


  it('should correctly match the snapshot', () => {
    wrapper = shallow(<Stats {...mockProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should have the proper default state', () => {
    const expectedState = { range: 'today' }
    expect(wrapper.state()).toEqual(expectedState)
  })

  it('should call changeFilter when the back button is clicked', () => {
    wrapper.find('.back-btn').simulate('click')
    expect(mockProps.changeFilter).toHaveBeenCalledWith('home')
  })

  describe('changeRange', () => {
    it('should correctly update the state when changeRange is called', () => {
      const mockEvent = { target: { value: 'all time' } }
      const mockEvent2 = { target: { value: 'today' } }
      const expectedState = { range: 'all time' }
      const expectedState2 = { range: 'today' }
      wrapper.instance().changeRange(mockEvent)
      expect(wrapper.state()).toEqual(expectedState)
      wrapper.instance().changeRange(mockEvent2)
      expect(wrapper.state()).toEqual(expectedState2)
    })

    it('should call handleData wiht getShotsAllTime if range is TODAY and changeRange is called' , () => {
      const mockEvent = { target: { value: 'all time' } }
      const mockUser = 1
      wrapper.setState({range: 'today'})
      wrapper.instance().changeRange(mockEvent)
      expect(mockProps.handleData).toHaveBeenCalledWith(getDataSuccess, getShotDataAllTime(mockUser))
    })

    it('should call handleData wiht getShotsToday if range is allTime and changeRange is called', () => {
      const mockEvent = { target: { value: 'today' } }
      const mockUser = 1
      wrapper.setState({ range: 'all time' })
      wrapper.instance().changeRange(mockEvent)
      expect(mockProps.handleData).toHaveBeenCalledWith(getDataSuccess, getShotDataToday(mockUser))
    })
  })


  describe('mapStateToProps', () => {
    it('should return an object with shotData and isLoading', () => {
      const mockState = {
        isLoading: false,
        shotData: [{ goodShots: 12 }],
        fakeState: 1,
        fakeState2: 'a'
      }

      const expected = {
        isLoading: false,
        shotData: [{ goodShots: 12 }],
      }

      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    it('should call dispatch with the handleData thunk', () => {
      const mockDispatch = jest.fn()
      const mockUser = 1
      const mappedProps = mapDispatchToProps(mockDispatch)
      const actionToDispatch = handleData(getDataSuccess, getShotDataAllTime(1))

      mappedProps.handleData(getDataSuccess, getShotDataAllTime(mockUser))
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})
