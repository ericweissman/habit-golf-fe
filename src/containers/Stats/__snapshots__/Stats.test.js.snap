// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stats should correctly match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Stats />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="date-btns"
        >
          <button
            onClick={[Function]}
            value="today"
          >
            today
          </button>
          <button
            onClick={[Function]}
            value="all time"
          >
            all time
          </button>
        </div>,
        <div
          className="graph-area"
        >
          <ReactMinimalPieChart
            animate={false}
            animationDuration={500}
            animationEasing="ease-out"
            className="graph"
            cx={50}
            cy={50}
            data={
              Array [
                Object {
                  "color": "#E38627",
                  "title": "One",
                  "value": 1.1,
                },
                Object {
                  "color": "#C13C37",
                  "title": "Two",
                  "value": 1.9,
                },
                Object {
                  "color": "#6A2135",
                  "title": "Three",
                  "value": 20,
                },
              ]
            }
            label={false}
            labelPosition={50}
            lengthAngle={360}
            lineWidth={100}
            paddingAngle={0}
            radius={50}
            ratio={1}
            rounded={false}
            startAngle={0}
          />
        </div>,
        <button />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              onClick={[Function]}
              value="today"
            >
              today
            </button>,
            <button
              onClick={[Function]}
              value="all time"
            >
              all time
            </button>,
          ],
          "className": "date-btns",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "today",
              "onClick": [Function],
              "value": "today",
            },
            "ref": null,
            "rendered": "today",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "all time",
              "onClick": [Function],
              "value": "all time",
            },
            "ref": null,
            "rendered": "all time",
            "type": "button",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <ReactMinimalPieChart
            animate={false}
            animationDuration={500}
            animationEasing="ease-out"
            className="graph"
            cx={50}
            cy={50}
            data={
              Array [
                Object {
                  "color": "#E38627",
                  "title": "One",
                  "value": 1.1,
                },
                Object {
                  "color": "#C13C37",
                  "title": "Two",
                  "value": 1.9,
                },
                Object {
                  "color": "#6A2135",
                  "title": "Three",
                  "value": 20,
                },
              ]
            }
            label={false}
            labelPosition={50}
            lengthAngle={360}
            lineWidth={100}
            paddingAngle={0}
            radius={50}
            ratio={1}
            rounded={false}
            startAngle={0}
          />,
          "className": "graph-area",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "animate": false,
            "animationDuration": 500,
            "animationEasing": "ease-out",
            "className": "graph",
            "cx": 50,
            "cy": 50,
            "data": Array [
              Object {
                "color": "#E38627",
                "title": "One",
                "value": 1.1,
              },
              Object {
                "color": "#C13C37",
                "title": "Two",
                "value": 1.9,
              },
              Object {
                "color": "#6A2135",
                "title": "Three",
                "value": 20,
              },
            ],
            "label": false,
            "labelPosition": 50,
            "lengthAngle": 360,
            "lineWidth": 100,
            "onClick": undefined,
            "onMouseOut": undefined,
            "onMouseOver": undefined,
            "paddingAngle": 0,
            "radius": 50,
            "ratio": 1,
            "rounded": false,
            "startAngle": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="date-btns"
          >
            <button
              onClick={[Function]}
              value="today"
            >
              today
            </button>
            <button
              onClick={[Function]}
              value="all time"
            >
              all time
            </button>
          </div>,
          <div
            className="graph-area"
          >
            <ReactMinimalPieChart
              animate={false}
              animationDuration={500}
              animationEasing="ease-out"
              className="graph"
              cx={50}
              cy={50}
              data={
                Array [
                  Object {
                    "color": "#E38627",
                    "title": "One",
                    "value": 1.1,
                  },
                  Object {
                    "color": "#C13C37",
                    "title": "Two",
                    "value": 1.9,
                  },
                  Object {
                    "color": "#6A2135",
                    "title": "Three",
                    "value": 20,
                  },
                ]
              }
              label={false}
              labelPosition={50}
              lengthAngle={360}
              lineWidth={100}
              paddingAngle={0}
              radius={50}
              ratio={1}
              rounded={false}
              startAngle={0}
            />
          </div>,
          <button />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                onClick={[Function]}
                value="today"
              >
                today
              </button>,
              <button
                onClick={[Function]}
                value="all time"
              >
                all time
              </button>,
            ],
            "className": "date-btns",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "today",
                "onClick": [Function],
                "value": "today",
              },
              "ref": null,
              "rendered": "today",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "all time",
                "onClick": [Function],
                "value": "all time",
              },
              "ref": null,
              "rendered": "all time",
              "type": "button",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <ReactMinimalPieChart
              animate={false}
              animationDuration={500}
              animationEasing="ease-out"
              className="graph"
              cx={50}
              cy={50}
              data={
                Array [
                  Object {
                    "color": "#E38627",
                    "title": "One",
                    "value": 1.1,
                  },
                  Object {
                    "color": "#C13C37",
                    "title": "Two",
                    "value": 1.9,
                  },
                  Object {
                    "color": "#6A2135",
                    "title": "Three",
                    "value": 20,
                  },
                ]
              }
              label={false}
              labelPosition={50}
              lengthAngle={360}
              lineWidth={100}
              paddingAngle={0}
              radius={50}
              ratio={1}
              rounded={false}
              startAngle={0}
            />,
            "className": "graph-area",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "animate": false,
              "animationDuration": 500,
              "animationEasing": "ease-out",
              "className": "graph",
              "cx": 50,
              "cy": 50,
              "data": Array [
                Object {
                  "color": "#E38627",
                  "title": "One",
                  "value": 1.1,
                },
                Object {
                  "color": "#C13C37",
                  "title": "Two",
                  "value": 1.9,
                },
                Object {
                  "color": "#6A2135",
                  "title": "Three",
                  "value": 20,
                },
              ],
              "label": false,
              "labelPosition": 50,
              "lengthAngle": 360,
              "lineWidth": 100,
              "onClick": undefined,
              "onMouseOut": undefined,
              "onMouseOver": undefined,
              "paddingAngle": 0,
              "radius": 50,
              "ratio": 1,
              "rounded": false,
              "startAngle": 0,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
